compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp

```diff
@@ -0,0 +1,90 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===--------------- InterchangeTransposeGenericOps.cpp -------------------===//
+//
+// Interchange loops in generic ops to make the transpose happen on the outputs
+// instead of inputs.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Dialect/Flow/Transforms/FusionUtils.h"
+#include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+struct TransposeGenericOpPattern : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (genericOp.getNumInputs() == 0 || genericOp.getNumOutputs() == 0)
+      return failure();
+
+    // check the generic op can be fused with the input operands.
+    auto indexingMap0 =
```

Comment:
We need to tighten this up. We need to check 
1) If there is already a producer of this operation where the producer is a linalg named op and the indexing map is identity. If that is the case, do nothing (cause the assumption is that it will fuse with that producer)
2) If not, pick the first named op producer and make the indexing map identity w.r.t that.